### PR TITLE
Use Toolchains to decouple Gradle build and test JDK

### PR DESCRIPTION
**Pull request by codecholeric**
This will introduce [Gradle Toolchains](https://docs.gradle.org/current/userguide/toolchains.html) into the ArchUnit build. The benefit is that we can decouple the JRE used to run Gradle from the JDK used to compile specific source sets and the JRE with which we run the test tasks. There was a lot of custom build script logic (`multiJdkTest` and similar) to facilitate running tests with different JREs, but one test gap was always that the code was compiled with the newest JDK necessary to support all features (at the moment JDK 15, limited by the need to be compatible with JDK 7).

With this PR the behavior is now the following:

* no extra option configured for the build -> set JVM compatibility to minimum (at the moment Java 7) and use the newest JDK possible to compile (at the moment JDK 15, since otherwise we run into problems to be compatible with Java 7)
* with the option `-PtestJavaVersion=x` the compile will run with JDK `x` wherever possible. If `x` is too low (e.g. source set needs JDK 9 but `x < 9`) it will use the minimum JDK compatible with the source set. Also tests will now be run with a JRE `x` or the minimum compatible JRE and the Maven integration test will be run with JDK `x`.

There were a couple of incompatibilities that only showed when really using and older JDK as opposed to just setting the target compatibility to a lower level and using a newer JDK.

Note that at the moment we can still not run tests compiled with a newer JDK version than Gradle 6 supports, because the Gradle test plugin does some introspection of the class files and crashes when it doesn't understand the class file format (supposedly to identify which classes pass to the JUnit runner). I.e. it then passes on wrong classes as test classes.

However, this will open up the way to upgrade Gradle even if while we still need to be Java 7 compatible.

Resolves: #477

### CODE REVIEW

File: archunit/src/main/java/com/tngtech/archunit/core/domain/AccessTarget.java [hunk 1]
```diff
@@ -440,8 +440,9 @@ public JavaClass getRawReturnType() {
         @Override
         @PublicAPI(usage = ACCESS)
         public ThrowsClause<CodeUnitCallTarget> getThrowsClause() {
+            @SuppressWarnings("RedundantTypeArguments") // Some JDK versions need the type argument, or it will not compile
             List<ThrowsClause<JavaCodeUnit>> resolvedThrowsClauses = FluentIterable.from(resolve())
-                    .transform(toGuava(JavaCodeUnit.Functions.Get.throwsClause()))
+                    .transform(toGuava(JavaCodeUnit.Functions.Get.<JavaCodeUnit>throwsClause()))
```
Comment (timtebeek) [inline archunit/src/main/java/com/tngtech/archunit/core/domain/AccessTarget.java:445]: Ah this [tripped me up as well](https://github.com/TNG/ArchUnit/pull/724/commits/3dd094d200ec257f17380bc3d77cc902eb274c5e); nice to see another way to resolve it!
